dialogy/types/entity/time_interval/__init__.py

```diff
@@ -150,6 +155,97 @@ def get_value(self) -> Any:
                 f"Expected at least 1 of `from` or `to` in {self.values} for {self}"
             )
 
+    @classmethod
+    def pick_value(
+        cls, d_values: List[Dict[str, Any]], grain: str, constraints: Dict[str, Any]
+    ) -> List[Dict[str, Any]]:
+        """
+        Filters interval datetime values outside of timerange constraint
+
+        Filter logic apply constraint on both "from" and "to" individually:
+        - both FROM and TO present and inside constraint
+            - no filtering
+        - both FROM and TO present but only FROM inside constraint
+            - filter TO
+        - both FROM and TO present but only TO inside constraint
+            - replace FROM value with constraint lowerbound datetime value
+        - both FROM and TO present but outside constraint
+            - filter both TO and FROM
+        - only FROM present and inside constraint
+            - no filtering
+        - only FROM present and outside constraint
+            - filter FROM
+
+        """
+        is_time = grain in const.TIME_UNITS
+        constraint = constraints.get(const.TIME)
+        if not is_time or not constraint:
+            return d_values
+
+        opt_items = [
+            (cls.__TIMERANGE_OPERATION_ALIAS.get(opt), measure)
+            for opt, measure in constraint.items()
+        ]
+
+        constrained_d_values = []
+        for datetime_val in d_values:
+            from_ = datetime_val.get(const.FROM) and all(
+                [
+                    opt(
+                        getattr(
+                            datetime.fromisoformat(
+                                datetime_val.get(const.FROM, {}).get(const.VALUE)
+                            ),
+                            const.HOUR,
+                        ),
+                        measure[const.HOUR],
+                    )
+                    for opt, measure in opt_items
+                    if opt
+                ]
+            )
+            to_ = datetime_val.get(const.TO) and all(
+                [
+                    opt(
+                        getattr(
+                            datetime.fromisoformat(
+                                datetime_val.get(const.TO, {}).get(const.VALUE)
+                            ),
+                            const.HOUR,
+                        ),
+                        measure[const.HOUR],
+                    )
+                    for opt, measure in opt_items
+                    if opt
+                ]
+            )
```

Comment:
`all(
                [
                    opt(
                        getattr(
                            datetime.fromisoformat(
                                datetime_val.get(const.FROM, {}).get(const.VALUE)
                            ),
                            const.HOUR,
                        ),
                        measure[const.HOUR],
                    )
                    for opt, measure in opt_items
                    if opt
                ]
            )`
            
two of these `all(...)` pieces should become a util function which will say return a bool or something similar or maybe both  `datetime_val.get(const.x) and all(...)` can be put into function with a proper name and params, as in current way it's unreadable and repetitive to me. Once made into functions they can be reused.